### ISSUE TITLE
JS Function calls take more time than nashorn

### ISSUE THREAD

**Issue by karthickpdy** [oracle/graal#799]
I am running benchmark between Nashorn and Graal. But I see that Nashorn is performing extremely better of the order of 40x w.r.t execution of javascript function. Can anybody explain me the same?

Have attached the benchmark file and the code file
[Multithread Function Benchmark.txt](https://github.com/oracle/graal/files/2569090/Multithread.Function.Benchmark.txt)
.
[FunctionBenchmark.txt](https://github.com/oracle/graal/files/2569086/FunctionBenchmark.txt)

**Comment by thomaswue** [comment]
You are running the Nashorn code in parallel in 30 threads, while you are synchronizing the Graal.js code to run sequentially. We published detailed performance results of Graal.js vs Nashorn on well-known benchmarks. See "https://medium.com/graalvm/oracle-graalvm-announces-support-for-nashorn-migration-c04810d75c1f".

**Comment by chumer** [comment]
You need to create a context with an explicit engine. This enables source caching between multiple contexts. A context for JavaScript can only be used from one thread at a time, but if you create a context for each thread with the same engine then your code will be shared and compiled only once.

More info here: http://www.graalvm.org/docs/graalvm-as-a-platform/embed/#enable-source-caching

**Comment by karthickpdy** [comment]
Thanks for the response. I tried out your suggestion and did a benchmark. But I do not see considerable performance differences between caching and not caching. What happens when we enable source caching? Is the whole AST shared? If so, why the performance in both cases are so similar. Here is the [gist](https://gist.github.com/karthickpdy/a37bd7644a2dabe7b8ad8034733de167) for reference. This is the [source](https://github.com/jashkenas/underscore/blob/master/underscore.js.) I used. Let me know if I am missing something.
